string uri
string username
string password
